plugin/restore/mysql/mysql.go

```diff
@@ -680,6 +680,181 @@ func (r *Restore) GetSortedBinlogFilesMetaOnServer(ctx context.Context) ([]Binlo
 	return sortBinlogFiles(binlogFiles), nil
 }
 
+// GetBinlogCoordinateByTs converts a timestamp to binlog coordinate using local binlog files.
+func (r *Restore) GetBinlogCoordinateByTs(ctx context.Context, targetTs int64) (*api.BinlogInfo, error) {
+	binlogFilesLocalSorted, err := GetSortedLocalBinlogFiles(r.binlogDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read sorted local binlog files, error: %w", err)
+	}
+	if !binlogFilesAreContinuous(binlogFilesLocalSorted) {
+		return nil, fmt.Errorf("local binlog files are not continuous")
+	}
+
+	var binlogFileTarget BinlogFile
+	var found bool
+	for i, file := range binlogFilesLocalSorted {
+		eventTs, err := r.parseFirstLocalBinlogEventTs(ctx, file.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse the local binlog file %q's first binlog event ts, error: %w", file.Name, err)
+		}
+		if eventTs >= targetTs {
+			if i == 0 {
+				return nil, fmt.Errorf("the targetTs %d is before the first event ts %d", targetTs, eventTs)
+			}
+			// The previous local binlog file contains targetTs.
+			found = true
+			binlogFileTarget = binlogFilesLocalSorted[i-1]
+			break
+		}
+	}
+	// All of the local binlog files' first event start ts <= targetTs, so we choose the last binlog file as "containing" targetTs.
+	// This may not be true, because possibly targetTs > last eventTs of the last binlog file.
+	// In this case, we think that the user want's to recover to the latest state of the database, and convert the ts to the last eventTs of the last binlog file.
+	if !found {
+		binlogFileTarget = binlogFilesLocalSorted[len(binlogFilesLocalSorted)-1]
+	}
+
+	eventPos, err := r.getBinlogEventPositionAtOrAfterTs(ctx, binlogFileTarget, targetTs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find the binlog event after targetTs %d, error: %w", targetTs, err)
+	}
+	return &api.BinlogInfo{FileName: binlogFileTarget.Name, Position: eventPos}, nil
+}
+
+func parseBinlogEventTsInLine(line string) (eventTs int64, found bool, err error) {
+	// The target line starts with string like "#220421 14:49:26 server id 1"
+	if !strings.Contains(line, "server id") {
+		return 0, false, nil
+	}
+	if strings.Contains(line, "end_log_pos 0") {
+		// https://github.com/mysql/mysql-server/blob/8.0/client/mysqlbinlog.cc#L1209-L1212
+		// Fake events with end_log_pos=0 could be generated and we need to ignore them.
+		return 0, false, nil
+	}
+	fields := strings.Fields(line)
+	// fields should starts with ["#220421", "14:49:26", "server", "id", "1", "end_log_pos", "34794"]
+	if len(fields) < 7 ||
+		(len(fields[0]) != 7 || len(fields[1]) != 8 || fields[2] != "server" || fields[3] != "id" || fields[5] != "end_log_pos") {
+		return 0, false, fmt.Errorf("found unexpected mysqlbinlog output line %q when parsing binlog event timestamp and end position", line)
```

Comment:
nit: there is a disparity between the error log and the method name.

I would remove the "and end position" part.

Reply:
Done.